README.md

```diff
@@ -2,51 +2,71 @@
+    - `task_id/constants.RESOURCES`
+    - `task_id/constants.NETWORK_RESOURCES`
+    - `task_id/constants.RESULTS`
+    - `task_id/constants.NETWORK_RESULTS`
+  - should treat `work_dir/task_id` as the working directory for the given task
+  - `task_params_json` is a JSON string containing task specific parameters
+  - will only be called once with given `task_id`
+  - can assume `task_id/constants.RESOURCES` contains all the resources provided by task creator
+- `NextSubtask`
+  - takes one argument `task_id`
+  - can assume `CreateTask` was called earlier with the same `task_id`
+  - returns `subtask_id` which can be an arbitrary string but the same string cannot be returned more than once
+  - also returns `subtask_params_json` which is the JSON string containing subtask specific parameters
+- `HasPendingSubtasks`
+  - takes one argument `task_id`
+  - returns a boolean indicating whether there are any more pending subtasks waiting for computation at given moment
+  - in case when it returns `true`, the next `NextSubtask` call should return successfully
+- `Verify`
+  - takes two arguments, `task_id` and `subtask_id` which specify which subtask results should be verified
+  - will be called with only valid `task_id` and `subtask_id` values
+  - returns a boolean indicating whether results passed the verification or not
+  - for successfully verified subtasks it most likely should also perform merging the partial results into the final one
+- `DiscardSubtasks`
+  - takes two arguments, `task_id` and `subtask_ids`
+  - should discard results of given subtasks and any dependent subtasks
+  - returns list of subtask IDs that have been discarded
+  - in a simple case where subtasks are independent from each other it will return the same list as it received
+- `Benchmark`
+  - takes no arguments
+  - returns a score which indicates how efficient the machine is for this type of tasks
+  - shouldn't take much time (preferably less than a minute for medium range machines)
+- `Shutdown`
+  - takes no arguments
+  - should gracefully terminate the service
 
 When the last subtask is successfully verified on the requestor's side, the `work_dir/task_id/constants.RESULTS` directory should contain all result files and nothing else.
+
+## Provider
+Provider side is a set of one-off commands. There are currently two commands it should support.
+
+- `Compute`
+  - gets a single working directory to operate on `task_work_dir`
+  - different subtasks of the same task will have the same `task_work_dir`
+  - takes `task_id`, `subtask_id`, `subtask_params_json` as arguments
+  - can assume the `task_work_dir/constants.NETWORK_RESOURCES` directory exists
+  - can assume that under `task_work_dir/constants.NETWORK_RESOURCES` are the resources specified in the corresponding `NextSubtask` call
+  - generates a single `result.zip` file under `task_work_dir/subtask_id/` (subject likely to change) which is the result that will be sent back to the requestor
```

Comment:
Do we need this constraint? Maybe Golem could just copy *all contents* from `task_work_dir/subtask_id/` on provider's machine to `task_work_dir/constants.NETWORK_RESULTS/subtask_id` on requestor's machine?

Reply:
I think it's a safe and nonlimiting constraint that the result should be a single file (easier to manage), but I was thinking that the `Compute` methods could return the path to the result.